docker/docker.go

```diff
@@ -138,96 +142,72 @@ func NewImage(conf *Config) (*Image, error) {
 		Transport: tr,
 		Timeout:   conf.Timeout,
 	}
-	registry := dockerHub
-	tag := "latest"
-	token := ""
-	os := "linux"
-	arch := "amd64"
-	var nameParts, tagParts []string
-	var name, port string
-	state := stateInitial
-	start := 0
-	for i, c := range conf.ImageName {
-		if c == ':' || c == '/' || c == '@' || i == len(conf.ImageName)-1 {
-			if i == len(conf.ImageName)-1 {
-				// ignore a separator, include the last symbol
-				i += 1
-			}
-			part := conf.ImageName[start:i]
-			start = i + 1
-			switch state {
-			case stateInitial:
-				if part == "localhost" || strings.Contains(part, ".") {
-					// it's registry, let's check what's next =port of image name
-					registry = part
-					if c == ':' {
-						state = statePort
-					} else {
-						state = stateName
-					}
-				} else {
-					// it's an image name, if separator is /
-					// next part is also part of the name
-					// othrewise it's an offcial image
-					if c == '/' {
-						// we got just a part of name, till next time
-						start = 0
-						state = stateName
-					} else {
-						state = stateTag
-						name = fmt.Sprintf("library/%s", part)
-					}
-				}
-			case stateTag:
-				tag = ""
-				tagParts = append(tagParts, part)
-			case statePort:
-				state = stateName
-				port = part
-			case stateName:
-				if c == ':' || c == '@' {
-					state = stateTag
-				}
-				nameParts = append(nameParts, part)
-			}
-		}
+
+	image := &Image{
+		user:     conf.User,
+		password: conf.Password,
+		Token:    "",
+		client:   client,
+		os:       "linux",
+		arch:     "amd64",
 	}
 
-	if port != "" {
-		registry = fmt.Sprintf("%s:%s", registry, port)
+	ref, err := reference.ParseNormalizedNamed(conf.ImageName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse image name. name=%v: %v", conf.ImageName, conf.ImageName)
 	}
-	if name == "" {
-		name = strings.Join(nameParts, "/")
+
+	// strip tag if image has digest and tag
+	ref = imageNameWithDigestOrTag(ref)
+	// add default tag "latest"
+	ref = reference.TagNameOnly(ref)
+
+	image.Registry = reference.Domain(ref)
+	// reference.ParseNormalizedNamed setting `defaultDomain`, we need `dockerHub` domain
+	if image.Registry == defaultDomain {
+		image.Registry = dockerHub
 	}
-	if tag == "" {
-		tag = strings.Join(tagParts, ":")
+	image.Name = reference.Path(ref)
+	if canonical, isDigested := ref.(reference.Canonical); isDigested {
+		image.Tag = canonical.Digest().String()
```

Comment:
Why a digest is put into a tag?

Reply:
this field is tag or digest - we used it the same to pull the layers from the repo.
URL = <registry><name>/manifests/<tag/digest>

Reply:
from https://docs.docker.com/registry/spec/api/
"GET /v2/<name>/manifests/<reference>"
I can rename field to `reference`